service/src/main/java/bio/terra/tanagra/service/authentication/JwtAccessTokenUtils.java

```diff
@@ -0,0 +1,95 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.EncodedKeySpec;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Base64;
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+import org.bouncycastle.util.io.pem.PemObject;
+import org.bouncycastle.util.io.pem.PemReader;
+import org.springframework.util.Assert;
+
+public class JwtAccessTokenUtils {
+  private final JWTVerifier jwtVerifier;
+  private final ObjectMapper mapper = new ObjectMapper();
+  private final Base64.Decoder decoder = Base64.getUrlDecoder();
+
+  private static byte[] parsePEMFile(String pemFileName) throws IOException {
+    File pemFile = FileUtils.getResourceFile(Path.of("keys/" + pemFileName));
+    PemReader reader = new PemReader(new FileReader(pemFile));
+    PemObject pemObject = reader.readPemObject();
+    byte[] content = pemObject.getContent();
+    reader.close();
+    return content;
+  }
+
+  private static PublicKey getPublicKey(byte[] keyBytes, String algorithm) {
+    try {
+      KeyFactory kf = KeyFactory.getInstance(algorithm);
+      EncodedKeySpec keySpec = new X509EncodedKeySpec(keyBytes);
+      return kf.generatePublic(keySpec);
+    } catch (NoSuchAlgorithmException e) {
+      throw new InvalidCredentialsException(
+          "Public key generation failed, algorithm not be found: " + algorithm, e);
+    } catch (InvalidKeySpecException e) {
+      throw new InvalidCredentialsException("Public key generation failed, invalid key", e);
+    }
+  }
+
+  private static PublicKey readPublicKeyFromFile(String publicKeyPemFileName, String algorithm)
+      throws IOException {
+    byte[] bytes = parsePEMFile(publicKeyPemFileName);
+    return getPublicKey(bytes, algorithm);
+  }
+
+  public JwtAccessTokenUtils(String issuer, String publicKeyPemFileName, String algorithmName)
+      throws IOException {
+    Assert.isTrue(StringUtils.isNotBlank(issuer), "JWT issuer empty");
+    Algorithm algorithm =
+        Algorithm.RSA256(
+            (RSAPublicKey) readPublicKeyFromFile(publicKeyPemFileName, algorithmName),
+            /* RSAPrivateKey= */ null);
+    jwtVerifier = JWT.require(algorithm).withIssuer(issuer).build();
+  }
+
+  public UserId getUserIdFromToken(String accessToken) {
+    Assert.isTrue(StringUtils.isNotBlank(accessToken), "user accessToken empty");
+
+    try {
+      DecodedJWT verifiedJWT = jwtVerifier.verify(accessToken);
+      String payloadJSON = new String(decoder.decode(verifiedJWT.getPayload()));
```

Comment:
We verify the JWT when Tanagra is behind IAP, but from the discussion in Slack, it's not clear to me that we should be doing the verifying here or whether the proxy in front will do that. Maybe better for a discussion over GVC?

Reply:
From discussion -
downstream VWB deployment is behind proxy, which takes care of authn. Services call sam to decode token.
verification was added for non-proxy deployments, and hence will be made optional here.